### PR TITLE
Segment.io Analytics

### PR DESCRIPTION
**Pull request by tonydaly**
This PR for #193 adds the segment.io tracking code and configures it to:

- [x] identify logged in users

And track key actions:
- [ ] Signed up
- [ ] Logged in
- [ ] Liked
- [x] Viewed Page

There is a mixture of client-side and server-side tracking. We use the `SEGMENT_API_KEY` environment variable to configure the account. We won't render the tracking partial if this key isn't present to avoid issues in development/staging environments.

### CODE REVIEW

**Comment by tonydaly** [comment]
This currently has the client-side tracking working fine, but I'm still working on the server-side tracking, using [analytics-elixir](https://github.com/stueccles/analytics-elixir).

**Comment by tonydaly** [comment]
# Update:

I've added a [behaviour](http://elixir-lang.org/getting-started/typespecs-and-behaviours.html) for our Analytics module. This allows us to have multiple implementations of the behaviours we set out. For example, this now has two implementations: `Local` and `Segment`

The local implementation is for use in development and testing where we don't want to be talking to an external API (and muddying up out data), it implements the `Analytics` behaviour, but only logs information. 

The second implementation talks directly to the Segment API via the [`analytics-elixir`](https://github.com/stueccles/analytics-elixir) library.

I've also added two shared modules `EventNameBuilder` and `PropertiesBuilder`, these should hopefully allow us to easily extend which events and properties we track across our different models. I hope adding new events is straight forward (pattern matching based on the struct and action), but I can add more documentation if needed. There's currently an example of tracking new Goals in the `GoalController` (`track_created`).

**Comment by tonydaly** [comment]
# Issue:

The implementation of `analytics-elixir` doesn't seem to be correctly sending the events. It is POSTing an event to Segment, but they aren't showing up in the debugger. Segment returns a 200 for everything, including [errors](analytics-elixir), which makes debugging this a real pain.

By curling with our write key I can create an event, so I suspect there's a bug in the library. I've started looking into it...

**Comment by tonydaly** [comment]
I've got a feeling that the issue is with the write key not being properly [encoded](https://segment.com/docs/sources/server/http/#authentication). Looking into the library atm...

**Comment by tonydaly** [comment]
@katbow there's still an issue in the Segment library, but I think the interface here is at the point where it can be reviewed. The idea behind all of this is that it abstracts the individual implementation details anyway.

File: config/config.exs [hunk 1]
```diff
@@ -22,6 +22,9 @@ config :logger, :console,
   format: "$time $metadata[$level] $message\n",
   metadata: [:request_id]
 
+config :segment,
+  write_key: System.get_env("SEGMENT_WRITE_KEY")
```
Comment (katbow) [inline config/config.exs]: @tonydaly can you add this to heroku if you haven't yet?

File: lib/healthlocker.ex [hunk 2]
```diff
@@ -14,6 +14,7 @@ defmodule Healthlocker do
       supervisor(Healthlocker.Endpoint, []),
       # Start your own worker by calling: Healthlocker.Worker.start_link(arg1, arg2, arg3)
       # worker(Healthlocker.Worker, [arg1, arg2, arg3]),
+      worker(Segment, [Application.get_env(:segment, :write_key)])
```
Comment (katbow) [inline lib/healthlocker.ex]: I came across `System.get_env` and `Application.get_env` when I was making the feedback feature. Can you explain the difference/when to use which? I didn't fully understand from the docs. I had to use System.get_env for what I was doing and couldn't get Application.get_env to work, so I wasn't sure if there was a specific case to use this.

File: web/templates/layout/_segment.html.eex [hunk 3]
```diff
@@ -0,0 +1,91 @@
+<script type="text/javascript">
```
Comment (katbow) [inline web/templates/layout/_segment.html.eex]: Is there a reason for this to be here specifically and not in `web/static/js/` as a .js file?

Comment (tonydaly) [inline web/templates/layout/_segment.html.eex]: The file does some interpolation to inject the `SEGMENT_WRITE_KEY`, I tried creating a `.js` file with the `.eex` extension but it didn't seem to be working. I think it's probably a configuration issue, maybe there's a better way?

**Comment by katbow** [comment]
thanks for merging @Cleop! After you merge things, be sure to delete the branch as it is not necessary to have here anymore & keeps the number of branches for the repo to a minimum 👍

File: lib/healthlocker.ex (hunk 2 above)
Comment (tonydaly) [inline lib/healthlocker.ex]: [`System.get_env`](https://hexdocs.pm/elixir/System.html#get_env/1) fetches an actual environment variable, which is managed by the OS process. 

[`Application.get_env`](https://hexdocs.pm/elixir/Application.html#get_env/2) is an Elixir helper which is basically just doing an internal key/value lookup. You can see this key is set in [`config.exs`](https://github.com/healthlocker/healthlocker/blob/master/config/config.exs#L37). 

The pattern seems to be, load all your environment variables in the config, but then access all your configuration through `Application`.

**Pull request merged**